Share code for buffer swapping and copy-forward between RemoteLayerWithInProcessRendering and RemoteImageBufferSet.

mattwoodrow opened the pull request:
#### 27846c09e4fee72c441c3d78c0c93ed53bae2b2e
<pre>
Share code for buffer swapping and copy-forward between RemoteLayerWithInProcessRendering and RemoteImageBufferSet.
<a href="https://bugs.webkit.org/show_bug.cgi?id=266201">https://bugs.webkit.org/show_bug.cgi?id=266201</a>
&lt;<a href="https://rdar.apple.com/119474592">rdar://119474592</a>&gt;

Reviewed by Dan Glastonbury.

This moves a bunch of code into a base class &apos;ImageBufferSet&apos;, where the in-process and remote versions can both use it.
I think we can go further here, including volatility, and ideally RemoteImageBufferSetProxy would be a subclass to.

* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::ensureBufferForDisplay):
(WebKit::RemoteImageBufferSet::prepareBufferForDisplay):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::paintContents):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::hasFrontBuffer const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::frontBufferMayBeVolatile const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::clearBackingStore):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::frontBufferHandle const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::createContextAndPaintContents):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::createFlushers):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::setBufferVolatile):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::prepareToDisplay):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::encodeBufferAndBackendInfos const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::dump const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::prepareBuffers): Deleted.
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::setFrontBufferNonVolatile): Deleted.
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::swapToValidFrontBuffer): Deleted.
* Source/WebKit/Shared/graphics/ImageBufferSet.cpp: Added.
(WebKit::ImageBufferSet::swapBuffersForDisplay):
(WebKit::ImageBufferSet::computePaintingRects):
(WebKit::ImageBufferSet::prepareBufferForDisplay):
* Source/WebKit/Shared/graphics/ImageBufferSet.h: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:

Canonical link: <a href="https://commits.webkit.org/283678@main">https://commits.webkit.org/283678@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3c322aa1600c9469659c52bf71af3c562d967d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67017 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46392 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19639 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71052 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18150 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54191 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17933 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53751 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12215 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70084 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42670 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57985 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34270 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39341 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16504 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61269 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15720 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72753 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10974 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15056 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61220 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11006 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58044 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61296 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14810 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9014 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2620 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42199 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43276 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44459 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43019 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 55;
+	objectVersion = 54;
```
:grrr:

djg commented on Source/WebKit/Shared/graphics/ImageBufferSet.h:
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
Checks ⌚ : It's 2024

djg commented on Source/WebKit/Shared/graphics/ImageBufferSet.cpp:
```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
Checks ⌚ : It's 2024

djg commented on Source/WebKit/Shared/graphics/ImageBufferSet.cpp:
```diff
@@ -0,0 +1,153 @@
+    if (m_frontBuffer) {
+        if (m_frontBuffer->setNonVolatile() == WebCore::SetNonVolatileResult::Empty)
+            m_previouslyPaintedRect = std::nullopt;
+    }
```
Why nested `if` and not `if (m_frontBuffer && m_frontBuffer->set[...])`?

djg commented on Source/WebKit/Shared/graphics/ImageBufferSet.cpp:
```diff
@@ -0,0 +1,153 @@
+    return displayRequirement;
+}
+
+Vector<FloatRect, 5> ImageBufferSet::computePaintingRects(const Region& dirtyRegion, float resolutionScale)
```
Should the magic number 5 be a named constant and/or Vector<FloatRect, 5> have a `using` alias? (eg. `using PaintRects = Vector<FloatRect, 5>`)

djg commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
```diff
@@ -52,22 +52,21 @@
-    m_secondaryBackBuffer.discard();
+    m_bufferSet.m_frontBuffer = nullptr;
+    m_bufferSet.m_backBuffer = nullptr;
+    m_bufferSet.m_secondaryBackBuffer = nullptr;
```
Is it future work to make this operation an API on the buffer set?

djg reviewed (approved):
:shipit:

mattwoodrow committed:
Share code for buffer swapping and copy-forward between RemoteLayerWithInProcessRendering and RemoteImageBufferSet.
https://bugs.webkit.org/show_bug.cgi?id=266201
<rdar://119474592>

Reviewed by Dan Glastonbury.

This moves a bunch of code into a base class 'ImageBufferSet', where the in-process and remote versions can both use it.
I think we can go further here, including volatility, and ideally RemoteImageBufferSetProxy would be a subclass to.

* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::ensureBufferForDisplay):
(WebKit::RemoteImageBufferSet::prepareBufferForDisplay):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::paintContents):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::hasFrontBuffer const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::frontBufferMayBeVolatile const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::clearBackingStore):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::frontBufferHandle const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::createContextAndPaintContents):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::createFlushers):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::setBufferVolatile):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::prepareToDisplay):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::encodeBufferAndBackendInfos const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::dump const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::prepareBuffers): Deleted.
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::setFrontBufferNonVolatile): Deleted.
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::swapToValidFrontBuffer): Deleted.
* Source/WebKit/Shared/graphics/ImageBufferSet.cpp: Added.
(WebKit::ImageBufferSet::swapBuffersForDisplay):
(WebKit::ImageBufferSet::computePaintingRects):
(WebKit::ImageBufferSet::prepareBufferForDisplay):
* Source/WebKit/Shared/graphics/ImageBufferSet.h: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:

Canonical link: https://commits.webkit.org/283678@main
